### PR TITLE
[Feature][Connector-V2] Support extract topic from SeaTunnelRow fields

### PR DESCRIPTION
**Pull request by TaoZex**
## Purpose of this pull request

Support extract topic from SeaTunnelRow fields，use value of a field from upstream data as topic,the format is ${field_topic}, where topic is the name of one of the columns of the upstream data.

## Check list

* [ ] Code changed are covered with tests, or it does not need tests for reason:
* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/incubator-seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/incubator-seatunnel/tree/dev/docs

### CODE REVIEW

File: docs/en/connector-v2/sink/Kafka.md [hunk 1]
```diff
@@ -31,6 +31,12 @@ By default, we will use 2pc to guarantee the message is sent to kafka exactly on
 
 Kafka Topic.
 
+Currently two formats are supported：
+
+1. Fill in the name of the topic
+
+2. Use value of a field from upstream data as topic,the format is ${field_topic}, where topic is the name of one of the columns of the upstream data.
```
Comment (hailin0) [inline docs/en/connector-v2/sink/Kafka.md:38]:
```suggestion
2. Use value of a field from upstream data as topic,the format is ${your field name}, where topic is the value of one of the columns of the upstream data.
```

File: seatunnel-connectors-v2/connector-kafka/src/main/java/org/apache/seatunnel/connectors/seatunnel/kafka/sink/KafkaSinkWriter.java [hunk 2]
```diff
@@ -86,6 +95,8 @@ public KafkaSinkWriter(
             List<KafkaSinkState> kafkaStates) {
         this.context = context;
         this.pluginConfig = pluginConfig;
+        this.topic = extractTopic(pluginConfig.getString(TOPIC));
+        this.topicExtractor = createTopicExtractor(this.topic, seaTunnelRowType);
```
Comment (hailin0) [inline seatunnel-connectors-v2/connector-kafka/src/main/java/org/apache/seatunnel/connectors/seatunnel/kafka/sink/KafkaSinkWriter.java:99]:
```suggestion
        this.topicExtractor = createTopicExtractor(pluginConfig.getString(TOPIC), seaTunnelRowType);
```

File: seatunnel-connectors-v2/connector-kafka/src/main/java/org/apache/seatunnel/connectors/seatunnel/kafka/sink/KafkaSinkWriter.java [hunk 3]
```diff
@@ -54,10 +56,13 @@ public class KafkaSinkWriter implements SinkWriter<SeaTunnelRow, KafkaCommitInfo
     private final SinkWriter.Context context;
     private final Config pluginConfig;
     private final Function<SeaTunnelRow, String> partitionExtractor;
+    private final Function<SeaTunnelRow, String> topicExtractor;
 
     private String transactionPrefix;
     private long lastCheckpointId = 0;
     private int partition;
+    private String topic;
+    private boolean isExtractTopic = false;
```
Comment (hailin0) [inline seatunnel-connectors-v2/connector-kafka/src/main/java/org/apache/seatunnel/connectors/seatunnel/kafka/sink/KafkaSinkWriter.java:65]: remove

File: seatunnel-connectors-v2/connector-kafka/src/main/java/org/apache/seatunnel/connectors/seatunnel/kafka/sink/KafkaSinkWriter.java [hunk 4]
```diff
@@ -188,6 +211,24 @@ private void restoreState(List<KafkaSinkState> states) {
         }
     }
 
+    private Function<SeaTunnelRow, String> createTopicExtractor(String topicField, SeaTunnelRowType seaTunnelRowType) {
+        if (!isExtractTopic) {
+            return row -> null;
+        }
```
Comment (hailin0) [inline seatunnel-connectors-v2/connector-kafka/src/main/java/org/apache/seatunnel/connectors/seatunnel/kafka/sink/KafkaSinkWriter.java:217]:
```suggestion
    private Function<SeaTunnelRow, String> createTopicExtractor(String topicConfig, SeaTunnelRowType seaTunnelRowType) {
        String regex = "\\$\\{(.*?)\\}";
        Pattern pattern = Pattern.compile(regex, Pattern.DOTALL);
        Matcher matcher = pattern.matcher(topicConfig);
        if (!matcher.find()) {
            return row -> topicConfig;
        }
        String topicField = matcher.group(1);
```

File: seatunnel-connectors-v2/connector-kafka/src/main/java/org/apache/seatunnel/connectors/seatunnel/kafka/sink/KafkaSinkWriter.java [hunk 5]
```diff
@@ -188,6 +211,24 @@ private void restoreState(List<KafkaSinkState> states) {
...
+        List<String> fieldNames = Arrays.asList(seaTunnelRowType.getFieldNames());
+        if (!fieldNames.contains(topicField)) {
+            throw new IllegalArgumentException("Field name is not found!");
```
Comment (hailin0) [inline seatunnel-connectors-v2/connector-kafka/src/main/java/org/apache/seatunnel/connectors/seatunnel/kafka/sink/KafkaSinkWriter.java:220]:
```suggestion
            throw new IllegalArgumentException("Field name{" +  topicField + "} is not found!");
```

File: seatunnel-connectors-v2/connector-kafka/src/main/java/org/apache/seatunnel/connectors/seatunnel/kafka/sink/KafkaSinkWriter.java [hunk 6]
```diff
@@ -188,6 +211,24 @@ private void restoreState(List<KafkaSinkState> states) {
...
+        }
+        int topicFieldIndex = seaTunnelRowType.indexOf(topicField);
+        return row -> {
+            Object topicFieldValue = row.getField(topicFieldIndex);
+            if (topicFieldValue == null) {
+                throw new IllegalArgumentException("The column value is empty!");
```
Comment (hailin0) [inline seatunnel-connectors-v2/connector-kafka/src/main/java/org/apache/seatunnel/connectors/seatunnel/kafka/sink/KafkaSinkWriter.java:226]:
```suggestion
                throw new IllegalArgumentException("The field{" + topicField + "} value is empty!");
```

**Review by hailin0 — COMMENTED** [review]
Can you update e2e test this features?

example:
```hcon
sink {
  Kafka {
    bootstrap.servers = "xxx:xxx"
    topic = "${your field name}"
  }
}
```

**Comment by iture123** [comment]
I suggest support mutile fields,and support combination of fixed string and field names, such as topic = "topic_${fieldName1}_${fieldName2}",reference https://github.com/apache/incubator-seatunnel/blob/dev/seatunnel-connectors-v2/connector-elasticsearch/src/main/java/org/apache/seatunnel/connectors/seatunnel/elasticsearch/serialize/index/IndexSerializerFactory.java

**Comment by TaoZex** [comment]
> I suggest support mutile fields,and support combination of fixed string and field names, such as topic = "topic_${fieldName1}_${fieldName2}",reference https://github.com/apache/incubator-seatunnel/blob/dev/seatunnel-connectors-v2/connector-elasticsearch/src/main/java/org/apache/seatunnel/connectors/seatunnel/elasticsearch/serialize/index/IndexSerializerFactory.java

Thanks for your advice.

**Pull request closed without merging**